lib/hooks.dart

```diff
@@ -0,0 +1,97 @@
+@JS()
+library hooks;
+
+import 'package:js/js.dart';
+import 'package:react/react.dart';
+
+@JS()
+abstract class React {
+  external static List<dynamic> useState(dynamic value);
+}
+
+/// The return value of [useState].
+///
+/// The current value of the state is available via [value] and
+/// functions to update it are available via [set] and [setTx].
+///
+/// Note are two rules for using Hooks (<https://reactjs.org/docs/hooks-rules.html>):
+///
+/// * Only call Hooks at the top level.
+/// * Only call Hooks from inside a [DartFunctionComponent].
+///
+/// Learn more: <https://reactjs.org/docs/hooks-state.html>.
+class StateHook<T> {
+  /// The first item of the pair returned by [React.useState].
+  T _value;
+
+  /// The second item in the pair returned by [React.useState].
+  void Function(dynamic) _setValue;
```

Comment:
You might be right... my brain hurts when thinking about Dart 2 function types haha, so I'm not sure.

Works for me so long as there are no type errors in DDC or dart2js!

Reply:
Changing this type breaks the tests for `setTx`. I think it's because `_setValue` is the updater function returned by the js useState and can take an argument of type `T newValue` or type `T computeNewValue(T oldValue)`.

Reply:
![3fzoer](https://user-images.githubusercontent.com/41018730/68604975-d4b59000-0468-11ea-9b41-9e03735c99b0.jpg)
https://github.com/dart-lang/sdk/issues/4938
https://github.com/dart-lang/language/issues/83